language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -211,60 +211,64 @@ typeOf (Constant _ (BuiltinName _ n)) = do
+        _             -> throwError (TypeMismatch x (void body) (TyFix () dummyTyName dummyType) nBodyTy)
+typeOf (Wrap x n ty body) = do
+    nBodyTy <- typeOf body
+    let fixed = tySubstitute (extractUnique n) (NormalizedType $ TyFix () (void n) (void ty)) (void $ NormalizedType ty)
     typeCheckStep
-    if tyReduce fixed == ty'
-        then pure (TyFix () (void n) (void ty))
-        else throwError (TypeMismatch x (void t) fixed (void ty'))
+    if tyReduce fixed == nBodyTy
+        then pure $ NormalizedType (TyFix () (void n) (void ty)) -- type annotations on terms must be normalized
+        else throwError (TypeMismatch x (void body) fixed nBodyTy)
 
 extractUnique :: TyNameWithKind a -> Unique
 extractUnique = nameUnique . unTyName . unTyNameWithKind
 
 -- TODO: make type substitutions occur in a state monad + benchmark
+-- | Substitute the given type into another type. The input types must be normalized - this prevents us creating redundant work. The
+-- output type is not normalized, since we may introduce new redexes.
 tySubstitute :: Unique -- ^ Unique associated with type variable
-             -> Type TyNameWithKind a -- ^ Type we are binding to free variable
-             -> Type TyNameWithKind a -- ^ Type we are substituting in
+             -> NormalizedType TyNameWithKind a -- ^ Type we are binding to free variable
+             -> NormalizedType TyNameWithKind a -- ^ Type we are substituting in
              -> Type TyNameWithKind a
-tySubstitute u ty = cata a where
+tySubstitute u (NormalizedType ty) = cata a . getNormalizedType where
     a (TyVarF _ (TyNameWithKind (TyName (Name _ _ u')))) | u == u' = ty
     a x                                                  = embed x
 
 -- also this should involve contexts
-tyReduce :: Type TyNameWithKind a -> Type TyNameWithKind a
-tyReduce (TyApp _ (TyLam _ (TyNameWithKind (TyName (Name _ _ u))) _ ty) ty') = tySubstitute u ty' (tyReduce ty) -- TODO: use the substitution monad here
-tyReduce (TyForall x tn k ty)                                                = TyForall x tn k (tyReduce ty)
-tyReduce (TyFun x ty ty') | isTypeValue ty                                   = TyFun x (tyReduce ty) (tyReduce ty')
-                          | otherwise                                        = TyFun x (tyReduce ty) ty'
-tyReduce (TyLam x tn k ty)                                                   = TyLam x tn k (tyReduce ty)
-tyReduce (TyApp x ty ty') | isTypeValue ty                                   = TyApp x (tyReduce ty) (tyReduce ty')
-                          | otherwise                                        = TyApp x (tyReduce ty) ty'
-tyReduce x                                                                   = x
+-- | Reduce any redexes inside a type.
+tyReduce :: Type TyNameWithKind a -> NormalizedType TyNameWithKind a
+-- TODO: is this case actually safe? Don't we need to reduce again after substituting?
```

Comment:
Yes, we need to reduce again. We may substitute a lambda (which is itself normalised) into a function position. This gives us a non-normalised type.

Reply:
I'm going to fix this separately if that's okay, let me see if I can construct a failing test case.